rfcs/20200810-tf-internal-api.md

```diff
@@ -0,0 +1,119 @@
+
+This means for any API that is added in 2.x release, it will remain the same in 
+2.x+1. If we choose to remove the same API, we will mark it as deprecated in 
+2.x+1, and delete it at 2.x+2 release. TensorFlow is released every 3 months, 
+and this will give enough time for all the clients to change to the new 
+API/alternative.
+
+Any deprecation and backward incompatible API change will be explicitly called 
+out in TF release notes. This applies to both core API and "internal" API.
+
+### Acceptance Criteria
+The candidate of the "internal" API should:
+
+1. Does NOT fit for core TF API, otherwise it should be exposed as core TF API.
+1. Are effectively required to build/test/deploy a framework or library on top 
+   of TensorFlow, or to integrate such a framework into TensorFlow's API.
+1. Mature enough and won't change behavior/interface for every release. 
+1. Non-trivial, otherwise the function should be copied to the client side.
+1. Widely used in the implementation of TF public APIs (i.e. new functionality 
+   isn’t immediately added to the tf.internal namespace)
+1. Has at least two downstream libraries which are known to need it.
+
+TF API owners will review the new API proposal, and follow the existing review 
+process for core TF API.
+
+
+### Documentation and Docstring
+The "internal" API should have the same style and standard as the core 
+TensorFlow API, which is documented [here](https://github.com/tensorflow/community/blob/master/governance/api-reviews.md#docstrings). 
+We should explicitly list out the difference between "internal" API and core
+API, and also choose a different place on tensorflow.org so that existing user 
+are not confused.
+
+### Naming and sub namespace
+Similar "internal" APIs should be grouped together as sub namespaces, e.g., test 
+related APIs should live under "tf.internal.test". This is aligned with the 
+existing TF naming conversion.
+
+Try not to export experimental APIs since the "internal" API should be mature 
+enough.
+
+### Current candidate
+The following list is created from the private TF method usage within Keras, 
+when we were trying to convert Keras to use the public TF API only. This is by 
+no means the full list, but will serve as the first wave of review requests we 
+send to the API owner for review. We don't expect all of them to be approved, 
+and will discuss with the API owner on a case to case basis.
+
+|Symbol location  |API Name  | 
+:-------------- |:---------------------------------------------------- |
+|python.framework.func_graph.FuncGraph |tf.internal.FuncGraph     |
+|python.framework.combinations.*| tf.internal.test.combinations.* |
+|python.distribute.strategy_combinations.* |tf.internal.test.combinations.* |
+|python.training.tracking.base.no_automatic_dependency_tracking|tf.internal.tracking.no_automatic_dependency_tracking |
+|python.util.object_identity.*|tf.internal.object_identity.* |
```

Comment:
object_identity.py doesn't depend on TF. Should it be a separate (pure Python) library that both TF and Keras use?

There's a more general question about whether that kind of dependency is OK to put in tf.internal due to the friction of splitting it out / packaging it / etc., balanced with the danger of tf.internal becoming a repository for small-ish nifty Python things that aren't otherwise TF related.

Reply:
Yes, it can, and I think it would be minimally troubling since it's a simple, rarely or never changing, pure Python package.

Reply:
> There's a more general question about whether that kind of dependency is OK to put in tf.internal due to the friction of splitting it out / packaging it / etc., balanced with the danger of tf.internal becoming a repository for small-ish nifty Python things that aren't otherwise TF related.

I have the same concern for the general topic.

Reply:
Yes, ideally this doesn't have TF specific logic and can be put into a standalone python package. Do we have any existing package that can host it (eg google absl?) or we need to create a new one?

Reply:
I don't think we should expose object_identity at all. Rather, we should push framework developers to use `ref` or just solve the problem independently. In particular, don't see why should make this API available to framework builders and not for end-users.

Reply:
Also I think object_identity is fairly trivial and thus I believe fails one of the acceptance criteria.